src/transforms/groupby.js

```diff
@@ -82,43 +90,68 @@ exports.transform = function(data, state) {
 
     var newData = [];
 
-    data.forEach(function(trace) {
-        newData = newData.concat(transformOne(trace, state));
+    data.forEach(function(trace, i) {
+
+        newData = newData.concat(transformOne(trace, state, state.attributeSets[i]));
     });
 
     return newData;
 };
 
-function transformOne(trace, state) {
+function initializeArray(newTrace, a) {
+    Lib.nestedProperty(newTrace, a).set([]);
+}
+
+function pasteArray(newTrace, trace, j, a) {
+    Lib.nestedProperty(newTrace, a).set(
+        Lib.nestedProperty(newTrace, a).get().concat([
+            Lib.nestedProperty(trace, a).get()[j]
+        ])
+    );
+}
+
+function transformOne(trace, state, attributeSet) {
+
     var opts = state.transform;
-    var groups = opts.groups;
+    var groups = trace.transforms[state.transformIndex].groups;
```

Comment:
@etpinard the latter is the pre-transform information. We want to obtain the groups. If a `filter` applies before a `groupby` then `state.transform.groups` will be an array that's already culled by the filter.